src/NHibernate/Intercept/AbstractFieldInterceptor.cs

```diff
@@ -164,13 +165,22 @@ private bool IsUninitializedProperty(string fieldName)
 			return uninitializedFields != null && uninitializedFields.Contains(fieldName);
 		}
 
-		private object InitializeOrGetAssociation(INHibernateProxy value, string fieldName)
+		private object InitializeOrGetAssociation(object target, INHibernateProxy value, string fieldName)
 		{
 			if(value.HibernateLazyInitializer.IsUninitialized)
 			{
 				value.HibernateLazyInitializer.Initialize();
 				value.HibernateLazyInitializer.Unwrap = true; // means that future Load/Get from the session will get the implementation
 				loadedUnwrapProxyFieldNames.Add(fieldName);
+				// Set the property value in order to be accessible when the session is closed
```

Comment:
I think this should be done out of the `if`. Otherwise, this test would fail:

```c#
[Test]
public void CanGetInitializedLazyManyToOneAfterClosedSession()
{
	Order order;
	Payment payment;

	using (var s = OpenSession())
	{
		payment = s.Load<Payment>(1);
		NHibernateUtil.Initialize(payment);
		order = s.Get<Order>(1);
		// Initialize Payment property
		payment = order.Payment;
	}

	Assert.That(order.Payment, Is.EqualTo(payment));
}
```

Maybe it should be put in directly in `Intercept`, in the `if (value.IsProxy() && IsUninitializedAssociation(fieldName))`.

Reply:
The assert won't fail as the objects are equal, but if the assert is modified to:
```
Assert.That(order.Payment is WireTransfer, Is.True);
```
then it would fail as the proxy was not unwrapped on get. By appling your suggestions the test will start working but there is another scenario that it would still fail. If in the above test the following line is removed:
```
// Initialize Payment property
payment = order.Payment;
```
the test would fail as the proxy was not unwrapped and set. This could be fixed by using the following code:
```
if (session != null || !value.IsProxy())
{
  loadedUnwrapProxyFieldNames.Add(fieldName);
}
else if (value is INHibernateProxy proxy)
{
  // Unwap the proxy if it is initialized
  var unwrappedProxy = UnwrapOrNull(proxy);
  if (unwrappedProxy != null)
  {
    loadedUnwrapProxyFieldNames.Add(fieldName);
    return unwrappedProxy;
  }
}
//...
private static object UnwrapOrNull(INHibernateProxy proxy)
{
  if (!proxy.HibernateLazyInitializer.IsUninitialized && proxy.HibernateLazyInitializer.Session != null)
  {
    return proxy.HibernateLazyInitializer.GetImplementation(proxy.HibernateLazyInitializer.Session);
  }

  return null;
}

```
The problem is that the current implementation for set interception looks like:
```
if (this.__fieldInterceptor != null)
{
  this.__fieldInterceptor.MarkDirty();
  this.__fieldInterceptor.Intercept(this, <ReflectHelper.GetPropertyName(setter)>, value, true);
}
base.<setter>(value);
``` 
which prevents us from overriding the value that will be set. If you agree I would change the implementation so that it would work similar like the current getter:
```
if (this.__fieldInterceptor != null)
{
  this.__fieldInterceptor.MarkDirty();
  var result = this.__fieldInterceptor.Intercept(this, <ReflectHelper.GetPropertyName(setter)>, value, true);
  if (result != AbstractFieldInterceptor.InvokeImplementation)
  {
    value = (<getter.ReturnType>)result;
  }
}
base.<setter>(value);
```
By doing this we would be able to set the unwrapped proxy when populating values in the 2-phase load and both test would pass.

Reply:
I do not think the second case should be supported. The lazy property has not been accessed at all before closing the session. Having a failure in such case, even if we could avoid it, is to be expected. Not failing in this second case could be a troublesome legacy for later changes.

Reply:
> I do not think the second case should be supported.

I am ok with the decision as if we would support the second case we would also need to support the case when the `no-proxy` property is initialized after the parent is loaded to be consistent, example:
```
payment = s.Load<Payment>(1);
order = s.Get<Order>(1);
NHibernateUtil.Initialize(payment);
```
Which would be harder to implement as the parent would need to "subscribe" to the persistence context in order to initialize the `no-proxy` property once it is loaded.